### DISCUSSION TITLE
Distributed MinIO cluster failed several weeks at the same point in time.

### DISCUSSION THREAD

**Discussion by ftx0day** [minio/minio#17311] [Q&A] [answered]
## Current Behavior
My baremetal distributed MinIO cluster failed several weeks at the same point in time(Every Monday at 0:00 sharp)
There is no error log for each node for a period of time starting from 0:00, but it has been 503 unreadable and writable externally.
## Possible Solution
**This is the solution I tried**
a、Restart the entire cluster and skip the 00.00 point in time
   //But it gets stuck, some nodes appear waiting to be unlocked,So I had to restart the nodes that were stuck.
     I observed that these nodes had plenty of free CPU/ memory, but in terms of experience, restarting Minio caused a lot of congestion.
b、Let it happened，and use ‘systemctl restart minio' restart all nodes in the same time，some nodes get stuck with this command, and restart these stucked nodes.

## Steps to Reproduce (for bugs)

1. Every Monday at 0:00 sharp，happen on time.
2. Restart the entire cluster at the same time every few days. Some nodes need to restart physical servers to recover the cluster.

**It takes about one minute to restart a physical machine normally, but it takes about 10 minutes to restart a physical machine with Minio stuck.**

## Context

Here are some records of the failures：
Aug 15 01:27:24 worker-node-16-68 minio[3453441]: Waiting for all MinIO sub-systems to be initialized.. trying to acquire lock
Aug 15 01:27:30 worker-node-16-68 minio[3453441]: Waiting for all MinIO sub-systems to be initialized.. trying to acquire lock
Aug 15 01:27:40 worker-node-16-68 minio[3453441]: Waiting for all MinIO sub-systems to be initialized.. trying to acquire lock
Aug 15 01:27:47 worker-node-16-68 minio[3453441]: Waiting for all MinIO sub-systems to be initialized.. trying to acquire lock
Aug 15 01:27:55 worker-node-16-68 minio[3453441]: Waiting for all MinIO sub-systems to be initialized.. trying to acquire lock
Aug 15 01:28:02 worker-node-16-68 minio[3453441]: Waiting for all MinIO sub-systems to be initialized.. trying to acquire lock
Aug 15 01:28:10 worker-node-16-68 minio[3453441]: Waiting for all MinIO sub-systems to be initialized.. trying to acquire lock
Aug 15 01:29:08 worker-node-16-68 minio[3453441]: Waiting for all MinIO sub-systems to be initialized.. trying to acquire lock
Aug 15 01:29:08 worker-node-16-68 minio[3453441]: Waiting for all MinIO sub-systems to be initialized.. trying to acquire lock
Aug 15 01:29:08 worker-node-16-68 minio[3453441]: Waiting for all MinIO sub-systems to be initialized.. trying to acquire lock
Aug 15 01:29:08 worker-node-16-68 minio[3453441]: Waiting for all MinIO sub-systems to be initialized.. trying to acquire lock
Aug 15 01:29:08 worker-node-16-68 minio[3453441]: Waiting for all MinIO sub-systems to be initialized.. trying to acquire lock
Aug 15 01:29:08 worker-node-16-68 minio[3453441]: Waiting for all MinIO sub-systems to be initialized.. trying to acquire lock
Aug 15 01:29:08 worker-node-16-68 minio[3453441]: Waiting for all MinIO sub-systems to be initialized.. trying to acquire lock
Aug 15 01:29:10 worker-node-16-68 minio[3453441]: Waiting for all MinIO sub-systems to be initialized.. trying to acquire lock
Aug 15 01:29:19 worker-node-16-68 minio[3453441]: Waiting for all MinIO sub-systems to be initialized.. trying to acquire lock
Aug 15 01:29:28 worker-node-16-68 minio[3453441]: Waiting for all MinIO sub-systems to be initialized.. trying to acquire lock
Aug 15 01:29:37 worker-node-16-68 minio[3453441]: Waiting for all MinIO sub-systems to be initialized.. trying to acquire lock

**Maybe about this:
https://github.com/minio/minio/commit/fe157166cae7ae505fd72cf07a37431decef1515**
## Your Environment

* Version used (`minio --version`): Version: 2022-02-12T00:51:25Z
* Server setup and configuration:  
Refer to the official documentation for SystemMD deployment:
https://docs.min.io/minio/baremetal/installation/deploy-minio-distributed.html
By 'ps -ef | grep minio' :
/usr/local/bin/minio server --console-address :9001 http://10.200.16.74:9000/minio-1 http://10.200.16.74:9000/minio-2 http://10.200.16.74:9000/minio-3 http://10.200.16.74:9000/minio-4 http://10.200.16.75:9000/minio-1 http://10.200.16.75:9000/minio-2 http://10.200.16.75:9000/minio-3 http://10.200.16.75:9000/minio-4 http://10.200.16.67:9000/minio-1 http://10.200.16.67:9000/minio-2 http://10.200.16.67:9000/minio-3 http://10.200.16.67:9000/minio-4 http://10.200.16.68:9000/minio-1 http://10.200.16.68:9000/minio-2 http://10.200.16.68:9000/minio-3 http://10.200.16.68:9000/minio-4

* Operating System and version (`uname -a`): Ubuntu 18.04/20.04
* Physical Machine configuration:  32C64GB 、10GbE LAN、SSD Disks

How can I avoid these problems. Perhaps run with containers instead of systemmd or others?  At least do not restart the physical machine..

**I would love to know the possible reasons. Any answers that may be helpful are appreciated. Thank you.**

**Comment by harshavardhana** [comment]
@ftx0day you need to upgrade.

**Comment by harshavardhana** [comment]
> * /usr/local/bin/minio server --console-address :9001 http://10.200.16.74:9000/minio-1 http://10.200.16.74:9000/minio-2 http://10.200.16.74:9000/minio-3 http://10.200.16.74:9000/minio-4 http://10.200.16.75:9000/minio-1 http://10.200.16.75:9000/minio-2 http://10.200.16.75:9000/minio-3 http://10.200.16.75:9000/minio-4 http://10.200.16.67:9000/minio-1 http://10.200.16.67:9000/minio-2 http://10.200.16.67:9000/minio-3 http://10.200.16.67:9000/minio-4 http://10.200.16.68:9000/minio-1 http://10.200.16.68:9000/minio-2 http://10.200.16.68:9000/minio-3 http://10.200.16.68:9000/minio-4
> *

This is also the incorrect legacy style of starting MinIO please follow the new guidelines on using ellipses, please re-do your setup.

**Comment by ftx0day** [comment]
> > * /usr/local/bin/minio server --console-address :9001 http://10.200.16.74:9000/minio-1 http://10.200.16.74:9000/minio-2 http://10.200.16.74:9000/minio-3 http://10.200.16.74:9000/minio-4 http://10.200.16.75:9000/minio-1 http://10.200.16.75:9000/minio-2 http://10.200.16.75:9000/minio-3 http://10.200.16.75:9000/minio-4 http://10.200.16.67:9000/minio-1 http://10.200.16.67:9000/minio-2 http://10.200.16.67:9000/minio-3 http://10.200.16.67:9000/minio-4 http://10.200.16.68:9000/minio-1 http://10.200.16.68:9000/minio-2 http://10.200.16.68:9000/minio-3 http://10.200.16.68:9000/minio-4
> 
> This is also the incorrect legacy style of starting MinIO please follow the new guidelines on using ellipses, please re-do your setup.

If I follow the new guidelines on using ellipses, will the data in it be affected? For example, if I change each IP node to a domain name node. @harshavardhana

**Comment by harshavardhana** [comment]
> If I follow the new guidelines on using ellipses, will the data in it be affected? For example, if I change each IP node to a domain name node.

Yes it will be affected, it is practically a new setup.

**Comment by ftx0day** [comment]
> > If I follow the new guidelines on using ellipses, will the data in it be affected? For example, if I change each IP node to a domain name node.
> 
> Yes it will be affected, it is practically a new setup.

Thank you, I will upgrade after verification.

**Comment by fig0** [comment]
We have had the same problem for months. We have therefore updated to version "2022-08-13T21-54-44Z" last Wednesday (2022/10/26) and unfortunately the problem still occurred last Monday (2022/10/31). It is very frustrating. We have also tried to validate all components in the chain (Keycloak, VMs, network) but there are no abnormalities. The only constant is that the problem occurs every Monday between 0:00 - 04:00 a.m. UTC and applications cannot use MinIO for uploads and downloads. After a few hours (without intervention) the MinIO cluster is usable again (from about 05:00 a.m. UTC). It always happens only on Mondays ...

### Our Environment

-  Version used (minio --version): Version: 2022-08-13T21-54-44Z
- Server setup and configuration:
   /opt/minio/minio server --address :9000 --certs-dir /opt/minio/CA --console-address :9001 https://host{1...8}/data/disk{1...8}
- Operating System and version (uname -a): Linux 4.18.0-372.19.1.el8_6.x86_64
- Virtual Machine configuration: 8 Servers with 6C32GB, 10GbE LAN, 8 Disks (SSD Disks) each


@harshavardhana : You wrote that an upgrade might help. To which version should we upgrade?
@ftx0day: Did you find a solution?

**Comment by harshavardhana** [comment]
> We have had the same problem for months. We have therefore updated to version "2022-08-13T21-54-44Z" last Wednesday (2022/10/26) and unfortunately the problem still occurred last Monday (2022/10/31). It is very frustrating. We have also tried to validate all components in the chain (Keycloak, VMs, network) but there are no abnormalities. The only constant is that the problem occurs every Monday between 0:00 - 04:00 a.m. UTC and applications cannot use MinIO for uploads and downloads. After a few hours (without intervention) the MinIO cluster is usable again (from about 05:00 a.m. UTC). It always happens only on Mondays ...

@fig0 if you have a critical setup you should be subscribing as a customer with us so that we can help you. Other than that there is hardly any information here for me to help you..

**Comment by ftx0day** [comment]
> We have had the same problem for months. We have therefore updated to version "2022-08-13T21-54-44Z" last Wednesday (2022/10/26) and unfortunately the problem still occurred last Monday (2022/10/31). It is very frustrating. We have also tried to validate all components in the chain (Keycloak, VMs, network) but there are no abnormalities. The only constant is that the problem occurs every Monday between 0:00 - 04:00 a.m. UTC and applications cannot use MinIO for uploads and downloads. After a few hours (without intervention) the MinIO cluster is usable again (from about 05:00 a.m. UTC). It always happens only on Mondays ...
> 
> ### Our Environment
> * Version used (minio --version): Version: 2022-08-13T21-54-44Z
> * Server setup and configuration:
>   /opt/minio/minio server --address :9000 --certs-dir /opt/minio/CA --console-address :9001 [https://host{1...8}/data/disk{1...8}](https://host%7B1...8%7D/data/disk%7B1...8%7D)
> * Operating System and version (uname -a): Linux 4.18.0-372.19.1.el8_6.x86_64
> * Virtual Machine configuration: 8 Servers with 6C32GB, 10GbE LAN, 8 Disks (SSD Disks) each
> 
> @harshavardhana : You wrote that an upgrade might help. To which version should we upgrade? @ftx0day: Did you find a solution?



> We have had the same problem for months. We have therefore updated to version "2022-08-13T21-54-44Z" last Wednesday (2022/10/26) and unfortunately the problem still occurred last Monday (2022/10/31). It is very frustrating. We have also tried to validate all components in the chain (Keycloak, VMs, network) but there are no abnormalities. The only constant is that the problem occurs every Monday between 0:00 - 04:00 a.m. UTC and applications cannot use MinIO for uploads and downloads. After a few hours (without intervention) the MinIO cluster is usable again (from about 05:00 a.m. UTC). It always happens only on Mondays ...
> 
> ### Our Environment
> * Version used (minio --version): Version: 2022-08-13T21-54-44Z
> * Server setup and configuration:
>   /opt/minio/minio server --address :9000 --certs-dir /opt/minio/CA --console-address :9001 [https://host{1...8}/data/disk{1...8}](https://host%7B1...8%7D/data/disk%7B1...8%7D)
> * Operating System and version (uname -a): Linux 4.18.0-372.19.1.el8_6.x86_64
> * Virtual Machine configuration: 8 Servers with 6C32GB, 10GbE LAN, 8 Disks (SSD Disks) each
> 
> @harshavardhana : You wrote that an upgrade might help. To which version should we upgrade? @ftx0day: Did you find a solution?

I'm sorry for the late reply. I have taken the following measures to solve the problem:
1.  Update all Ubuntu 18.04LTS nodes to Centos 8
2.  Change 1 pool to 2 pools 
3.  Update Minio version to RELEASE.2022-08-13T21-54-44Z
During this period, you can set up a standby cluster to provide services
Personal suspicion is a problem with the system and hardware combination layer.

**Comment by fig0** [comment]
@ftx0day: Thanks a lot for your reply. I am happy to hear that you have found a solution.

In the meantime we did a lot of research and analysis and we were at least able to reproduce the error on our test environment. So, we now have the error always and only on Mondays in our test **and** production environment ...

We did the following steps in our test environment, consisting of 4 nodes with 1 disk each.

1. Stop all MinIO nodes
2. Wipe all disks
3. Change "MINIO_VOLUMES" from legacy syntax "https://host1:9000/data/disk1 https://host2:9000/data/disk1 https://host3:9000/data/disk1 https://host4:9000/data/disk1" to ellipses syntax https://host{1...4}:9000/data/disk1
4. Restart all MinIO nodes and restore bucket data via mc mirror cmd 

So basically, the only change in the setup is to use the ellipses syntax and since this change, we have this "Monday" issue in our test environment too. 

@harshavardhana: Can you please explain the semantic difference between the two syntax provided in step nr 3? For us, there should not be any difference, because even with the legacy syntax we had HA during rolling updates or maintenance. I also read a lot of your comments regarding the switch to ellipses syntax and performing a re-setup. We did all that, but we are still facing the issue ... Would it be a solution to switch both environments back to the legacy syntax (for the purpose of a robust and stable cluster)?

As a side note: We also have a third environment (QA, consisting of 4 nodes with 4 disk each), which still uses the legacy syntax for MINIO_VOLUMES. This environment is still not facing the issue - and all 3 environments (test, QA, production) are in the same network and installed with the same ansible playbook.

**Comment by fig0** [comment, marked as answer]
@ftx0day @harshavardhana: We have finally found the reason for this recurring issue. It was related to a default operating setting on RHEL. The execution of the command "fstrim" was scheduled on Mondays between 00:00-02:00 UTC on every MinIO server in the cluster. This results in a high CPU and IO load on each server, causing the MinIO nodes to mark several disks temporarily as offline ... 

You can verify this with the following command:
$ systemctl list-timers

NEXT                         LEFT          LAST                         PASSED       UNIT                         ACTIVATES
…
Mon 2023-03-06 01:33:25 UTC  6 days left   Mon 2023-02-27 00:33:02 UTC  8h ago       fstrim.timer                 fstrim.service


Our solution: we updated the schedule of fstrim to distribute the execution over the week.
